File: 1399.count-largest-group.js
/*
 * @lc app=leetcode id=1399 lang=javascript
 *
 * [1399] Count Largest Group
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var countLargestGroup = function(n) {
  let groups = []
  let sizes = Array(100).fill(0)
  for (let i=1;i<=n;i++){
    let sum = sumDigit(i)
    sizes[sum]++

    if (!groups[sum]) groups[sum] = []
    groups[sum].push(i)
  }
  sizes.sort((a,b)=>b-a) // could find max index instead
  console.log(sizes)
  return sizes.filter(v=>v===sizes[0]).length
};

function sumDigit(digit) {
  return (digit+'').split('').reduce((a,b)=>+b+a,0)
}
// @lc code=end

File: 1399.count-largest-group.py
#
# @lc app=leetcode id=1399 lang=python3
#
# [1399] Count Largest Group
#

# @lc code=start
import functools
class Solution:
    def countLargestGroup(self, n: int) -> int:
        hash = {}
        top = 0
        for i in range(1, n+1):
            arr = list(str(i))
            sum = functools.reduce(lambda a,v: a+int(v), arr, 0)
            if sum not in hash: hash[sum] = 0
            hash[sum] += 1
            if hash[sum] > top: top = hash[sum]

        return len([x for x in list(hash.values()) if x == top])

# @lc code=end

